Use schema_translate_map in get_lastrowid

Hi,

we have an issue when we use autoincrement PKs in combination with `schema_translate_map` set in the engine.
For example:
```python
metadata = MetaData()
my_table = Table(
    "my_table",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("name", String(1000), nullable=False),
    schema="my_schema"
)

engine = create_engine(
        conn_string,
        execution_options={"schema_translate_map": {"my_schema": "my_tenant_schema"}},
    )
conn = engine.connect()
conn.execute(my_table.insert().values(name="foo))
```
throws this error:
```Python Traceback
/caches/pypoetry/virtualenvs/all-KwnZMQER-py3.6/lib/python3.6/site-packages/sqlalchemy/engine/default.py:1273: in _setup_ins_pk_from_lastrowid
    lastrowid = self.get_lastrowid()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <sqlalchemy_exasol.base.EXAExecutionContext object at 0x7f695ec97eb8>

    def get_lastrowid(self):
        columns = self.compiled.sql_compiler.statement.table.columns
        autoinc_pk_columns = \
            [c.name for c in columns if c.autoincrement and c.primary_key]
        if len(autoinc_pk_columns) == 0:
            return None
        elif len(autoinc_pk_columns) > 1:
            util.warn("Table with more than one autoincrement, primary key" \
                      " Column!")
            raise Exception
        else:
            id_col = self.dialect.denormalize_name(autoinc_pk_columns[0])
    
            table = self.compiled.sql_compiler.statement.table.name
            table = self.dialect.denormalize_name(table)
    
            sql_stmnt = "SELECT column_identity from SYS.EXA_ALL_COLUMNS " \
                        "WHERE column_object_type = 'TABLE' and column_table " \
                        "= ? AND column_name = ?"
    
            schema = self.compiled.sql_compiler.statement.table.schema
            if schema is not None:
                schema = self.dialect.denormalize_name(schema)
                sql_stmnt += " AND column_schema = ?"
            cursor = self.create_cursor()
            if schema:
                cursor.execute(sql_stmnt, (table, id_col, schema))
            else:
                cursor.execute(sql_stmnt, (table, id_col))
            result = cursor.fetchone()
            cursor.close()
>           return int(result[0]) - 1
E           TypeError: 'NoneType' object is not subscriptable

/caches/pypoetry/virtualenvs/all-KwnZMQER-py3.6/lib/python3.6/site-packages/sqlalchemy_exasol/base.py:337: TypeError
```
in version 2.1.

For our particular case, the error can be circumvented by using the schema in the `schema_translate_map` in the error throwing `get_lastrowid` function:
```python
            if schema is not None:
                schema = self.compiled.sql_compiler.schema_translate_map.map_[schema] if schema in self.compiled.sql_compiler.schema_translate_map.map_ else schema # <--- explicitely use the translation map here 
                schema = self.dialect.denormalize_name(schema)
                sql_stmnt += " AND column_schema = ?"
```

--- a/sqlalchemy_exasol/base.py
+++ b/sqlalchemy_exasol/base.py
@@ -327,6 +327,7 @@ class EXAExecutionContext(default.DefaultExecutionContext):
 
             schema = self.compiled.sql_compiler.statement.table.schema
             if schema is not None:
+                schema = self.compiled.sql_compiler.schema_translate_map.map_[schema] if schema in self.compiled.sql_compiler.schema_translate_map.map_ else schema
                 schema = self.dialect.denormalize_name(schema)
                 sql_stmnt += " AND column_schema = ?"
 

--- a/sqlalchemy_exasol/base.py
+++ b/sqlalchemy_exasol/base.py
@@ -307,37 +307,42 @@ class EXAExecutionContext(default.DefaultExecutionContext):
 
     def get_lastrowid(self):
         columns = self.compiled.sql_compiler.statement.table.columns
-        autoinc_pk_columns = \
-            [c.name for c in columns if c.autoincrement and c.primary_key]
-        if len(autoinc_pk_columns) == 0:
+        autoinc_pk_columns = [c.name for c in columns if c.autoincrement and c.primary_key]
+        if not autoinc_pk_columns:
             return None
-        elif len(autoinc_pk_columns) > 1:
-            util.warn("Table with more than one autoincrement, primary key" \
-                      " Column!")
-            raise Exception
-        else:
-            id_col = self.dialect.denormalize_name(autoinc_pk_columns[0])
+        if len(autoinc_pk_columns) > 1:
+            msg = "Table with more than one autoincrement, primary key Column!"
+            util.warn(msg)
+            raise Exception(msg)
 
-            table = self.compiled.sql_compiler.statement.table.name
-            table = self.dialect.denormalize_name(table)
+        id_col = self.dialect.denormalize_name(autoinc_pk_columns[0])
+        table = self.compiled.sql_compiler.statement.table.name
+        table = self.dialect.denormalize_name(table)
 
-            sql_stmnt = "SELECT column_identity from SYS.EXA_ALL_COLUMNS " \
-                        "WHERE column_object_type = 'TABLE' and column_table " \
-                        "= ? AND column_name = ?"
+        def _schema(sql_compiler, dialect):
+            """Get the schema while taking the translation-map and the de-normalization into account"""
+            translate_map = sql_compiler.schema_translate_map
+            schema_dispatcher = dict(translate_map.map_) if translate_map else {}
+            schema = sql_compiler.statement.table.schema
+            schema = schema_dispatcher[schema] if schema in schema_dispatcher else schema
+            return dialect.denormalize_name(schema)
 
-            schema = self.compiled.sql_compiler.statement.table.schema
-            if schema is not None:
-                schema = self.compiled.sql_compiler.schema_translate_map.map_[schema] if schema in self.compiled.sql_compiler.schema_translate_map.map_ else schema
-                schema = self.dialect.denormalize_name(schema)
-                sql_stmnt += " AND column_schema = ?"
+        schema = _schema(
+            self.compiled.sql_compiler,
+            self.dialect
+        )
 
-            cursor = self.create_cursor()
-            if schema:
-                cursor.execute(sql_stmnt, (table, id_col, schema))
-            else:
-                cursor.execute(sql_stmnt, (table, id_col))
+        sql_stmnt = " ".join((
+            "SELECT column_identity from SYS.EXA_ALL_COLUMNS",
+            "WHERE column_object_type = 'TABLE' and column_table",
+            "= ? AND column_name = ?",
+            "AND column_schema = ?" if schema else ""
+        ))
+        args = (table, id_col, schema) if schema else (table, id_col)
+
+        with self.create_cursor() as cursor:
+            cursor.execute(sql_stmnt, args)
             result = cursor.fetchone()
-            cursor.close()
             return int(result[0]) - 1
 
     def pre_exec(self):

--- a/sqlalchemy_exasol/base.py
+++ b/sqlalchemy_exasol/base.py
@@ -319,15 +319,15 @@ class EXAExecutionContext(default.DefaultExecutionContext):
         table = self.compiled.sql_compiler.statement.table.name
         table = self.dialect.denormalize_name(table)
 
-        def _schema(sql_compiler, dialect):
+        def _get_schema(sql_compiler, dialect):
             """Get the schema while taking the translation-map and the de-normalization into account"""
             translate_map = sql_compiler.schema_translate_map
-            schema_dispatcher = dict(translate_map.map_) if translate_map else {}
+            schema_dispatcher = translate_map.map_ if translate_map else {}
             schema = sql_compiler.statement.table.schema
             schema = schema_dispatcher[schema] if schema in schema_dispatcher else schema
             return dialect.denormalize_name(schema)
 
-        schema = _schema(
+        schema = _get_schema(
             self.compiled.sql_compiler,
             self.dialect
         )
@@ -335,9 +335,9 @@ class EXAExecutionContext(default.DefaultExecutionContext):
         sql_stmnt = " ".join((
             "SELECT column_identity from SYS.EXA_ALL_COLUMNS",
             "WHERE column_object_type = 'TABLE' and column_table",
-            "= ? AND column_name = ?",
-            "AND column_schema = ?" if schema else ""
+            "= ? AND column_name = ?"
         ))
+        sql_stmnt += " AND column_schema = ?" if schema else ""
         args = (table, id_col, schema) if schema else (table, id_col)
 
         with self.create_cursor() as cursor:

--- a/test/test_regression.py
+++ b/test/test_regression.py
@@ -1,0 +1,42 @@
+"""This module contains various regression test for issues which have been fixed in the past"""
+from sqlalchemy import Column, Integer, MetaData, String, Table, create_engine, schema
+from sqlalchemy.testing import fixtures
+from sqlalchemy.testing.fixtures import config
+
+
+class RegressionTest(fixtures.TestBase):
+    def setUp(self):
+        self.table_name = "my_table"
+        self.tenant_schema_name = "tenant_schema"
+        engine = config.db
+        with config.db.connect() as conn:
+            conn.execute(schema.CreateSchema(self.tenant_schema_name))
+            metadata = MetaData()
+            Table(
+                self.table_name,
+                metadata,
+                Column("id", Integer, primary_key=True),
+                Column("name", String(1000), nullable=False),
+                schema=self.tenant_schema_name,
+            )
+            metadata.create_all(engine)
+
+    def tearDown(self):
+        with config.db.connect() as conn:
+            conn.execute(schema.DropSchema(self.tenant_schema_name, cascade=True))
+
+    def test_use_schema_translate_map_in_get_last_row_id(self):
+        """See also: https://github.com/exasol/sqlalchemy-exasol/issues/104"""
+        schema_name = "some_none_existent_schema"
+        options = {"schema_translate_map": {schema_name: self.tenant_schema_name}}
+        metadata = MetaData()
+        my_table = Table(
+            self.table_name,
+            metadata,
+            Column("id", Integer, primary_key=True),
+            Column("name", String(1000), nullable=False),
+            schema=schema_name,
+        )
+        engine = create_engine(config.db.url, execution_options=options)
+        with engine.connect() as conn:
+            conn.execute(my_table.insert().values(name="John Doe"))

